### ISSUE TITLE
Question about EnergyNet and ScoreNet

### ISSUE THREAD

**Issue by sseunghyuns** [Jiyao06/GenPose#25]
I'm curious if I understood the learning process of EnergyNet correctly. 
 
 Since it uses the same model architecture as ScoreNet, the input and output seem to be the same. 
 The loss also seems to be implemented similarly, calculating the l2 loss with the target_score (gt noise) for both. 
 
1. **The only difference seems to be that when calculating the energy, it takes the dot product of the pose and the output of EnergyNet. Is my understanding correct?**
 
2. **If so, it seems like either EnergyNet or ScoreNet alone could potentially solve the pose sampling and ranking problem. 
I'm wondering if there's something I'm missing!**

---

Additionally, in the paper, it mentioned that Energy-based model is parameterized as bellow:

<img width="560" alt="image" src="https://github.com/Jiyao06/GenPose/assets/63924704/9122369e-8386-4c63-8851-024207d2c5ea">

3. **Could you explain how the calculation came out? I couldn't find the answer even after referring to the relevant paper, so I'm asking for clarification.**

Thanks!

**Comment by Jiyao06** [comment]
Yes, your understanding of the network is correct. While using ScoreNet alone could obtain energy through integration, the diffusion model involves multiple steps of sampling, making this integration process time-consuming. Hence, we opted to separately train an EnergyNet. As for the reparameterization of EnergyNet, it aims to enhance training efficiency. You can find more insights in [1].

[1]Salimans, Tim, and Jonathan Ho. "Should EBMs model the energy or the score?." Energy Based Models Workshop-ICLR 2021. 2021.

**Comment by sseunghyuns** [comment]
Thanks for your answer! 

I have follow-up questions about the training EnergyNet:



1. I'm just double-checking to make sure I understand correctly. **Please let me know if I've misunderstood anything!**:

The score of a pose $\mathbf{p}$ (data point) is defined as the gradient of the log pdf  $\log p(\mathbf{p})$ with respect to $\mathbf{p}$: 

$$
\begin{align} s(\mathbf{p}) = \nabla_{\mathbf{p}} \log p(\mathbf{p}) \end{align}
$$

Given that the pdf in EBM can be expressed in terms of an energy function $E(\mathbf{p})$: ($Z$ is a normalization constant)

$$
\begin{align} p(\mathbf{p}) = \frac{e^{-E(\mathbf{p})}}{Z} \end{align}
$$

Thus we can derive the score as the negative gradient of the energy function with respect to the pose $\mathbf{p}$:

$$
\begin{align} \log p(\mathbf{p}) = \log (\frac{e^{-E(\mathbf{p})}}{Z}) = -E(\mathbf{p}) - \log Z  \newline  s(\mathbf{p}) = \nabla_{\mathbf{p}} \log p(\mathbf{p}) = -\nabla_{\mathbf{p}}E(\mathbf{p}) \end{align}
$$

Therefore, we can define EnergyNet’s objective function as (6), utilizing the same architecture as ScoreNet. 


<img width="574" alt="image" src="https://github.com/Jiyao06/GenPose/assets/63924704/728c3afe-39b0-4bb5-8fae-34f79ed3f44a">


---

2. As implemented in the code, updating the energy model $\nabla_{\mathbf{p}}\Psi_{\phi}(\mathbf{p},t|O)$ means that the model learns to make the gradient of the inner product between the pose and the score-shaped vector,$\langle \mathbf{p}, \Phi_{\phi}(p,t|O) \rangle$, point towards the ground truth noise $\nabla_{\mathbf{p}}\log p_t(\mathbf{p}|O)$. Intuitively, since we are training with gradient descending, can we say that the update is made in the direction that minimizes the value of the inner product, which is the energy?

https://github.com/Jiyao06/GenPose/blob/1c815e7e971e6c81e2eb0789483bf62ae845a9cb/networks/gf_algorithms/energynet.py#L210-L215

---

3. While training EnergyNet, did you utilize the output of the pre-trained ScoreNet? Or did you train ScoreNet and EnergyNet separately from scratch?

---

Thank you so much for your time and patience in answering my questions. Your responses have been incredibly helpful, and I really appreciate your willingness to share your knowledge.
